planner/core/rule_build_key_info.go

```diff
@@ -91,6 +94,25 @@ func (p *LogicalSelection) checkMaxOneRowCond(unique expression.Expression, cons
 	return okCorCol
 }
 
+// isUnique checks the unique index whose NotNullFlag is false since equivalence conditions can filter out null values,
+// so a unique index without a NotNullFlag can return at most one row.
+func isUnique(p *LogicalSelection, col *expression.Column) bool {
+	for _, child := range p.children {
+		ds, ok := child.(*DataSource)
+		if !ok {
+			continue
+		}
+		tblInfo := ds.tableInfo
+		for _, idx := range tblInfo.Indices {
+			listColOrigName := strings.Split(col.OrigName, ".")
+			if idx.Unique && len(idx.Columns) == 1 && idx.Columns[0].Name.L == listColOrigName[len(listColOrigName)-1:][0] {
```

Comment:
This check is too tricky and ad-hoc, what if the DataSource is not the direct child of the Selection?

Besides, the filter of the Selection is not checked at all, what if the filter is `col IS NULL`? 

A more general approach would be using `isNullRejected()` to check the filters. Before that, maintain the unique keys(which may contain duplicate NULLs) from bottom up, i.e, relaxes the check in `checkIndexCanBeKey` to save those nullable unique keys into another field. Then utilize these nullable unique keys and the filter to infer new MaxOneRow info if possible.

Reply:
@eurekaka I have modified the code, could you please have a look?